src/NHibernate.Test/Linq/NullComparisonTests.cs

```diff
@@ -17,6 +18,285 @@ public class NullComparisonTests : LinqTestCase
+			Expect(q, Does.Not.Contain("Input is null").IgnoreCase.And.Contain("Output is null").IgnoreCase, BothDifferent, InputSet, BothNull);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.RelatedItems.All(r => r.Output != o.Input));
+			Expect(q, Does.Not.Contain("Input is null").IgnoreCase.And.Contain("Output is null").IgnoreCase, InputSet, OutputSet, BothDifferent, BothNull);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.RelatedItems.All(r => r.Output != null && r.Output != o.Input));
+			Expect(q, Does.Not.Contain("Input is null").IgnoreCase.And.Not.Contain("Output is null").IgnoreCase, BothDifferent, OutputSet);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => (o.NullableOutput + o.Output) != o.Output);
+			ExpectAll(q, Does.Contain("is null").IgnoreCase);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => (o.Input + o.Output) != o.Output);
+			Expect(q, Does.Not.Contain("is null").IgnoreCase, BothSame, BothDifferent);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.Address.Street != o.Output);
+			Expect(q, Does.Contain("Input is null").IgnoreCase, BothDifferent, OutputSet, BothNull);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.Address.City != o.Output);
+			Expect(q, Does.Contain("Output is null").IgnoreCase, InputSet, BothNull);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.Address.City != null && o.Address.City != o.Output);
+			Expect(q, Does.Not.Contain("Output is null").IgnoreCase);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.Address.Street != null && o.Address.Street != o.NullableOutput);
+			Expect(q, Does.Contain("Output is null").IgnoreCase, InputSet, BothDifferent);
+		}
+
+		[Test]
+		public void NullInequalityWithNotNullSubSelect()
+		{
+			if (!Dialect.SupportsScalarSubSelects)
+			{
+				Assert.Ignore("Dialect does not support scalar subselects");
+			}
+
+			var q = session.Query<AnotherEntityRequired>().Where(o => o.RelatedItems.Count != 1);
+			Expect(q, Does.Not.Contain("is null").IgnoreCase);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.RelatedItems.Max(r => r.Id) != 0);
+			ExpectAll(q, Does.Contain("is null").IgnoreCase);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.RelatedItems.All(r => r.Output != null) != o.NullableBool);
+			ExpectAll(q, Does.Not.Contain("or case").IgnoreCase);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.RelatedItems.Where(r => r.Id == 0).Sum(r => r.Input.Length) != 5);
+			ExpectAll(q, Does.Contain("or (").IgnoreCase);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.RelatedItems.All(r => r.Output != null) != (o.NullableOutput.Length > 0));
+			Expect(q, Does.Not.Contain("or case").IgnoreCase);
+		}
+
+		[Test]
+		public void NullEqualityWithNotNull()
+		{
+			var q = session.Query<AnotherEntityRequired>().Where(o => o.Input == null);
+			Expect(q, Does.Not.Contain("or is null").IgnoreCase, OutputSet, BothNull);
```

Comment:
Can we ever have `or is null` ? Shouldn't we have an expression between `or` and `is null`?

Reply:
In this case the `Input` property is mapped as not null so the `is null` is not needed. The `or is null` is added when comparing a nullable property. I will add additional tests for nullable properties.

Reply:
My point is `or is null` is not a valid SQL expression, so I doubt it can ever be added into the generated query. So I do not think this assert can fail, be it with master state or with this PR state.

Reply:
You are right, I didn't recognized that. Now it is fixed.